### PR TITLE
Copy the spark-specific implementation of murmur32 hash from cudf into spark-rapids-jni

### PR DESCRIPTION
**Pull request by nvdbaranec**
This duplicates the implementation, cpp tests and java tests for the spark-specific murmur32 hash done in cudf by @rwlee.  The jni bindings now point to this implementation instead of cudf so in theory we could deprecate what's left in cudf.

One thing I didn't do was trim out the "Spark" name prefixes scattered around the code. Since the code is properly in Spark-land now there's no real need to be so verbose.  If people want, I can clean it up.

Includes a small refactoring of the java decimal128 conversion code (see `to_java_bigdecimal` in `hash.cuh`) which will also be used by xxhash64.

<s>Dependent on https://github.com/NVIDIA/spark-rapids-jni/pull/1244</s>

### CODE REVIEW

**Review by jlowe — COMMENTED** [review]
Marking this as a draft since it depends upon #1244.

This needs a signoff commit per the contributor guidelines.

File: src/main/cpp/tests/hash.cpp
Comment (rwlee) [inline src/main/cpp/tests/hash.cpp:1]: Is this just https://github.com/rapidsai/cudf/blob/branch-23.08/cpp/tests/hashing/hash_test.cpp minus the md5 tests? It looks like it but I wanted to double check

File: src/test/java/com/nvidia/spark/rapids/jni/HashTest.java
Comment (rwlee) [inline src/test/java/com/nvidia/spark/rapids/jni/HashTest.java:1]: Similarly, is this the relevant tests pulled from https://github.com/rapidsai/cudf/blob/branch-23.08/java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java?

File: src/main/cpp/src/hash.cuh [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+/**
+ * @brief Converts a cudf decimal128 value to a java bigdecimal value.
+ *
+ * @param key The cudf decimal value
+ *
+ * @returns A 128 bit value containing the converted decimal bits and a length
+ *          representing the relevant number of bytes in the value.
+ *
+ */
+__device__ __inline__ std::pair<__int128_t, cudf::size_type> to_java_bigdecimal(
```
Comment (rwlee) [inline src/main/cpp/src/hash.cuh:37]: Seems like it might belong with decimal utils now that this is being pulled out, but it may be too hashing specific.

**Review by rwlee — DISMISSED** [review]
Looks good to me, code is largely identical to the cudf code with minimal refactoring. Only concern is if the `to_java_bigdecimal` belongs with the decimal utils, but I think that's a nit.

File: src/main/cpp/src/HashJni.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+#include "cudf_jni_apis.hpp"
+#include "dtype_utils.hpp"
+#include "jni_utils.hpp"
+
+#include "hash.cuh"
+
+extern "C" {
+
+JNIEXPORT jlong JNICALL Java_com_nvidia_spark_rapids_jni_Hash_murmurHash32(
+  JNIEnv* env, jobject j_object, jint seed, jlongArray column_handles)
```
Comment (jlowe) [inline src/main/cpp/src/HashJni.cpp:26]:
```suggestion
  JNIEnv* env, jclass, jint seed, jlongArray column_handles)
```

File: src/main/cpp/src/HashJni.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+{
+  JNI_NULL_CHECK(env, column_handles, "array of column handles is null", 0);
+
+  try {
```
Comment (jlowe) [inline src/main/cpp/src/HashJni.cpp:30]:
```suggestion
  try {
    cudf::jni::auto_set_device(env);
```

File: src/main/java/com/nvidia/spark/rapids/jni/Hash.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
```
Comment (jlowe) [inline src/main/java/com/nvidia/spark/rapids/jni/Hash.java:2]:
```suggestion
 * Copyright (c) 2023, NVIDIA CORPORATION.
```

File: src/main/cpp/src/hash.cuh (hunk 1 above)
Comment (nvdbaranec) [reply src/main/cpp/src/hash.cuh:37]: Hmm.  Does this code do anything that you would not normally do if just converting from cudf decimal 128 -> java?  I don't think so, but you're the original author :)  

The only real danger with putting it in decimal utils is if someone pokes around in here later one and breaks hashing.  I'm not sure how likely that is.  @jlowe  any opinions?

File: src/main/cpp/tests/hash.cpp
Comment (nvdbaranec) [reply src/main/cpp/tests/hash.cpp:1]: Yes.

File: src/test/java/com/nvidia/spark/rapids/jni/HashTest.java
Comment (nvdbaranec) [reply src/test/java/com/nvidia/spark/rapids/jni/HashTest.java:1]: Yes.

**Review by jlowe — APPROVED** [review]

**Comment by nvdbaranec** [comment]
build

**Comment by abellina** [comment]
build

**Review by rwlee — APPROVED** [review]

**Comment by nvdbaranec** [comment]
merge

**Pull request merged**